src/collective/communicator.cc

```diff
@@ -12,14 +13,10 @@
 namespace xgboost {
 namespace collective {
 
-thread_local std::unique_ptr<Communicator> Communicator::communicator_{};
+thread_local std::unique_ptr<Communicator> Communicator::communicator_{new NoOpCommunicator()};
```

Comment:
See other comments on the no-op, I have some concerns that we will use this accidentally, we have issues where XGBoost failed to establish a working communicator group but proceed with distributed training without explicit error. I haven't been able to tackle those issues due to the complicated network setup others use. But still, it's something that we should have in mind.

Reply:
On this line https://github.com/dmlc/xgboost/blob/master/rabit/src/engine.cc#L73, rabit provides a default engine if it's not initialized. We have some code that depends on this behavior. The `NoOpCommunicator` is one way to replicate this behavior.

I think the user has to tell us if they are doing distributed training, either by entering the `CommunicatorContext`, or calling `Init` directly. Otherwise we'd have no way of knowing, and can't just return a `nullptr` in case they are not in the distributed mode.